common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+)
+
+var (
+	errUploadNonRetriable = errors.New("upload non-retriable error")
+)
+
+const (
+	// URIScheme is the scheme for the gcloud storage implementation
+	URIScheme = "gs"
+
+	targetHistoryBlobSize = 2 * 1024 * 1024 // 2MB
+	errEncodeHistory      = "failed to encode history batches"
+	errBucketHistory      = "failed to get google storage bucket handle"
+	errWriteFile          = "failed to write history to google storage"
+)
+
+type historyArchiver struct {
+	container     *archiver.HistoryBootstrapContainer
+	gcloudStorage connector.Client
+
+	// only set in test code
+	historyIterator archiver.HistoryIterator
+}
+
+type progress struct {
+	CurrentPageNumber int64
+	IteratorState     []byte
+}
+
+type getHistoryToken struct {
+	CloseFailoverVersion int64
+	HighestPart          int64
+	CurrentPart          int64
+	NextBatchIdx         int
+}
+
+// NewHistoryArchiver creates a new gcloud storage HistoryArchiver
+func NewHistoryArchiver(
+	container *archiver.HistoryBootstrapContainer,
+	config *config.GstorageArchiver,
+) (archiver.HistoryArchiver, error) {
+	storage, err := connector.NewClient(context.Background(), config)
+	if err == nil {
+		return newHistoryArchiver(container, nil, storage), nil
+	}
+	return nil, err
+}
+
+func newHistoryArchiver(container *archiver.HistoryBootstrapContainer, historyIterator archiver.HistoryIterator, storage connector.Client) archiver.HistoryArchiver {
+	return &historyArchiver{
+		container:       container,
+		gcloudStorage:   storage,
+		historyIterator: historyIterator,
+	}
+}
+
+// Archive is used to archive a workflow history. When the context expires the method should stop trying to archive.
+// Implementors are free to archive however they want, including implementing retries of sub-operations. The URI defines
+// the resource that histories should be archived into. The implementor gets to determine how to interpret the URI.
+// The Archive method may or may not be automatically retried by the caller. The ArchiveOptions are used
+// to interact with these retries including giving the implementor the ability to cancel retries and record progress
+// between retry attempts.
+// This method will be invoked after a workflow passes its retention period.
+func (h *historyArchiver) Archive(ctx context.Context, URI archiver.URI, request *archiver.ArchiveHistoryRequest, opts ...archiver.ArchiveOption) (err error) {
+	scope := h.container.MetricsClient.Scope(metrics.HistoryArchiverScope, metrics.DomainTag(request.DomainName))
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer func() {
+		sw.Stop()
+		if err != nil {
+
+			if err.Error() == errUploadNonRetriable.Error() {
+				scope.IncCounter(metrics.HistoryArchiverArchiveNonRetryableErrorCount)
+			}
+
+			if featureCatalog.NonRetriableError != nil {
+				err = featureCatalog.NonRetriableError()
+			}
+
```

Comment:
If it's a transient error, emit `HistoryArchiverArchiveTransientErrorCount` and return the error directly.

Reply:
Please also update this part.

Reply:
done